pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/MissingOverrideRule.java

```diff
@@ -0,0 +1,363 @@
+
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.Stack;
+
+import net.sourceforge.pmd.lang.java.ast.ASTAllocationExpression;
+import net.sourceforge.pmd.lang.java.ast.ASTAnnotation;
+import net.sourceforge.pmd.lang.java.ast.ASTClassOrInterfaceDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTCompilationUnit;
+import net.sourceforge.pmd.lang.java.ast.ASTEnumConstant;
+import net.sourceforge.pmd.lang.java.ast.ASTEnumDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTFormalParameter;
+import net.sourceforge.pmd.lang.java.ast.ASTFormalParameters;
+import net.sourceforge.pmd.lang.java.ast.ASTMethodDeclaration;
+import net.sourceforge.pmd.lang.java.rule.AbstractJavaRule;
+
+
+/**
+ * Flags missing @Override annotations.
+ *
+ * @author Clément Fournier
+ * @since 6.1.0
+ */
+public class MissingOverrideRule extends AbstractJavaRule {
+
+    private final Stack<MethodLookup> currentLookup = new Stack<>();
+
+
+    @Override
+    public Object visit(ASTCompilationUnit node, Object data) {
+        currentLookup.clear();
+        return super.visit(node, data);
+    }
+
+
+    @Override
+    public Object visit(ASTClassOrInterfaceDeclaration node, Object data) {
+        currentLookup.push(getMethodLookup(node.getType()));
+        super.visit(node, data);
+        currentLookup.pop();
+
+        return data;
+    }
+
+    @Override
+    public Object visit(ASTEnumDeclaration node, Object data) {
+        currentLookup.push(getMethodLookup(node.getType()));
+        super.visit(node, data);
+        currentLookup.pop();
+
+        return data;
+    }
+
+
+
+    @Override
+    public Object visit(ASTAllocationExpression node, Object data) {
+        if (node.isAnonymousClass()) {
+            currentLookup.push(getMethodLookup(node.getType()));
+        }
+        super.visit(node, data);
+
+        if (node.isAnonymousClass()) {
+            currentLookup.pop();
+        }
+
+        return data;
+    }
+
+
+    @Override
+    public Object visit(ASTEnumConstant node, Object data) {
+        // FIXME, ASTEnumConstant needs typeres support!
+        //        if (node.isAnonymousClass()) {
+        //            currentExploredClass.push(node.getType());
+        //        }
+        super.visit(node, data);
+
+        //        if (node.isAnonymousClass()) {
+        //            currentExploredClass.pop();
+        //        }
+
+        return data;
+    }
+
+
+    /**
+     * Returns a map of method name to methods with the same name (overloads).
+     * The map contains a MethodWrapper for all methods declared in this class.
+     *
+     * @param exploredType Type to explore
+     */
+    private MethodLookup getMethodLookup(Class<?> exploredType) {
+        if (exploredType == null) {
+            return null;
```

Comment:
If we return here an "empty" MethodLookup instance, we could avoid the null check for currentLookup.peek() later on.

Reply:
Yes, but that null check allows to fail early, before we checked all annotations and evaluated the arguments to `isOverridden`. The performance loss would probably be insignificant, but the null check, if it's written explicitly, also documents that the lookup can be null/ empty/ unresolved, whereas using an empty instance would hide it I think